### ISSUE TITLE
Bug: Transcription result has non-linear time sequence for transcript words

### ISSUE THREAD

**Issue by ronen** [lowerquality/gentle#107]
# TL;dr:

The TODO note at https://github.com/lowerquality/gentle/blob/master/gentle/transcription.py#L52 which says  `# Combine chunks / # TODO: remove overlap? ...or just let the sequence aligner deal with it`  is correct I think:  Leaving it to the sequence aligner to deal with leads to weird results, in some weird cases.

I'm happy to code up a PR to remove the overlap.  If you have any thoughts on how best to do that, let me know!

# Bug description

I've attached a zip file with audio and transcript:  [rita_pierson.zip](https://github.com/lowerquality/gentle/files/472329/rita_pierson.zip)  (FYI they're from [this TED talk](http://www.ted.com/talks/rita_pierson_every_kid_needs_a_champion))

There's a section of the transcript which has:

    ... so we could show everybody else how to do it."
    One of the students said, "Really?"
    (Laughter)
    I said, "Really. We have to show the other classes
    how to do it ...

When I run Gentle, the resulting transcribed words have timing glitches.  Most noticeably they're not in linearly increasing order! Also, there's an inappropriately long gap between "students" and "said".  Here's an exceprt of the data:

start |  case | word
----- |  ----- | -----
211.73 | success | the
211.84 | success | students
215.74 | success | said
216.16 | success | Really
None | not-found-in-audio | Laughter
None | not-found-in-audio | I
216 | success | said
216.17 | success | Really
216.64 | success | We
216.86 | success | have

## Analysis...

This is due to a perfect storm of the transcript having repeated words combined with misidentified words and not-found-in-audio words, all in the overlap region of two chunks, causing the `diff_align` result to end up putting together a bit of a frankenstein creation.  Here's the details:

There are two overlapping chunks:

* One chunk spans t=198 to t=218, covering this part of the transcript:

    * **transcript:** `... One of the students said, "Really?" (Laughter) I said, "Really. We have to show the other`

    * **kaldi finds]** `... one of the students they pay me you like i said really we have to show the other`

    Notice that kaldi misidentified `said, Really? (Laughter)` as `they pay me you like`.  

* The other chunk spans t=216 to t=236, covering this part of the transcript:

    * **transcript:** `"said, "Really. We have to show the other classes how to do it..."`

    * **kaldi finds:** `said really we have to show the other classes how to do it...`

    Here kaldi has identified all the words correctly.

Concatenating the chunk results gives:

 >    ... one of the students they pay me you like i said really we have to  show the other said really we have to show the other classes how to do  it ...

And because `"said really"` appears twice in the concatenation, `diff_align` makes it work by using both occurrences:

>    ... one of the students ~~they pay me you like i~~ said really ~~we have to show the other~~ [Laughter i] said really we have to show the other   classes now to do it...

Since both occurrences of `"said really"` were at about t=216, that's what they're both listed as in the final result, which means they end up overlapping temporally. That also explains the timing gap between `"students"` and `"said"`, since it deleted a few seconds worth of words.

(Also, along the way it dropped the `"i"` found in the first chunk, and brought it back from the transcript, causing `"i"` to be listed as out-of-transcript even though it was actually be identified properly.)

## What to be done about it?

As per the comment at https://github.com/lowerquality/gentle/blob/master/gentle/transcription.py#L52, it seems likely that the answer is to remove the overlap rather than letting the diff_aligner deal with it.

Removing the overlap without introducing new bugs could be slightly tricky:

* The overlapped words don't have the exact same timings

* What if the overlapped regions didn't find the exact same words?

I'll put some thought into how to deal with those robustly and try to code up a PR, but as I said above, If you have any suggestions, I'm all ears!

Cheers

**Comment by ronen** [comment]
After a little bit of thought, removing the overlap turned out to be pretty easy!  PR on its way...

**Comment by pietrop** [comment]
Do you know if this bug is present in the `0.9.1` os x desktop app release?
